File: Advanced Programming Languages/Assignment3/employee_schedule.cpp
#include <algorithm>
#include <cstdlib>
#include <iostream>
#include <map>
#include <random>
#include <set>
#include <string>
#include <vector>
#include <cctype>

namespace
{

const std::vector<std::string> DAYS = {
    "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"
};
const std::vector<std::string> SHIFTS = { "morning", "afternoon", "evening" };
const int MIN_EMPLOYEES_PER_SHIFT = 2;
const int MAX_DAYS_PER_EMPLOYEE = 5;
const int PREFERRED_SHIFT_CAP = 3;

// Define a alias for a 2D vector of strings. Its just easier than defining a class/struct.
using Schedule = std::vector<std::map<std::string, std::vector<std::string>>>;

// Compiled release builds exit after returning 0 from main(). This helps with showing what was there before exiting.
void CheckExitCommand(const std::string& input)
{
    std::string lower;
    for (unsigned char c : input) lower += static_cast<char>(std::tolower(c));
    if (lower == "exit")
    {
        std::cout << "Exiting.\n";
        std::exit(0);
    }
}

void WaitForEnterBeforeExit()
{
    std::cout << "\nPress Enter to exit...";
    std::cin.clear();
    std::cin.ignore(10000, '\n');
    std::cin.get();
}

void GetShiftPriorityInput(std::vector<std::string>& order)
{
    std::cout << "  Enter priority order (1=most preferred, 2=second, 3=least preferred):\n";
    order.resize(3);
    std::vector<bool> used(4, false);
    for (size_t index = 0; index < 3; ++index)
    {
        while (true)
        {
            std::cout << "    " << SHIFTS[index] << ": rank (1-3): ";
            std::string rank_str;
            if (!(std::cin >> rank_str))
            {
                std::cin.clear();
                std::cin.ignore(10000, '\n');
                continue;
            }
            CheckExitCommand(rank_str);
            int rank = 0;
            try { rank = std::stoi(rank_str); } catch (...) { rank = 0; }
            if (rank >= 1 && rank <= 3 && !used[rank])
            {
                used[rank] = true;
                order[rank - 1] = SHIFTS[index];
                break;
            }
            std::cout << "Invalid or duplicate rank. Use 1, 2, 3 each once.\n";
        }
    }
}

void CollectEmployeeData(
    std::map<std::string, std::vector<std::string>>& preferences,
    std::map<std::string, std::vector<std::string>>* priorityOrders,
    bool usePriority)
{
    std::cout << "Number of employees: ";
    std::string num_str;
    std::cin >> num_str;
    CheckExitCommand(num_str);
    int number_of_employees = 0;
    try { number_of_employees = std::stoi(num_str); } catch (...) {}
    if (number_of_employees <= 0) return;
    for (int index = 0; index < number_of_employees; ++index)
    {
        std::string name;
        std::cout << "Employee name: ";
        std::cin >> name;
        CheckExitCommand(name);
        if (name.empty()) continue;

        if (usePriority && priorityOrders)
        {
            std::vector<std::string> order;
            GetShiftPriorityInput(order);
            (*priorityOrders)[name] = std::move(order);
        }

        std::vector<std::string> day_preferences;
        for (const auto& day : DAYS)
        {
            while (true)
            {
                std::cout << "  " << day << " preferred shift (morning/afternoon/evening): ";
                std::string preference;
                std::cin >> preference;
                CheckExitCommand(preference);
                for (auto& character : preference) character = static_cast<char>(std::tolower(static_cast<unsigned char>(character)));
                if (preference == "morning" || preference == "afternoon" || preference == "evening")
                {
                    day_preferences.push_back(preference);
                    break;
                }
                std::cout << "Invalid. Enter morning, afternoon, or evening.\n";
            }
        }
        preferences[name] = std::move(day_preferences);
    }
}

bool CanAssign(const std::string& employee, int day,
               const Schedule& schedule,
               const std::vector<std::set<std::string>>& assignedToday,
               const std::map<std::string, int>& daysWorked)
{
    if (assignedToday[day].count(employee)) return false;
    auto iterator = daysWorked.find(employee);
    if (iterator != daysWorked.end() && iterator->second >= MAX_DAYS_PER_EMPLOYEE) return false;
    return true;
}

Schedule BuildSchedule(
    const std::map<std::string, std::vector<std::string>>& preferences,
    const std::map<std::string, std::vector<std::string>>* priorityOrders)
{
    Schedule schedule(7);
    for (int day_index = 0; day_index < 7; ++day_index)
    {
        for (const auto& shift_name : SHIFTS)
        {
            schedule[day_index][shift_name] = {};
        }
    }

    std::map<std::string, int> daysWorked;
    std::vector<std::set<std::string>> assignedToday(7);

    std::vector<std::string> employees;
    for (const auto& pair : preferences) employees.push_back(pair.first);
    std::random_device random_seed_device;
    std::mt19937 random_generator(random_seed_device());
    std::shuffle(employees.begin(), employees.end(), random_generator);

    // First pass: assign by preference in rounds; at most one day per employee per round.
    // Use 3 rounds so backfill has enough capacity to ensure at least 2 per shift on all days.
    std::set<std::string> assignedThisRound;
    for (int round_index = 0; round_index < 3; ++round_index)
    {
        assignedThisRound.clear();
        for (size_t index = 0; index < employees.size(); ++index)
        {
            const std::string& employee = employees[index];
            if (assignedThisRound.count(employee)) continue;
            if (daysWorked[employee] >= MAX_DAYS_PER_EMPLOYEE) continue;
            for (int day_offset = 0; day_offset < 7; ++day_offset)
            {
                int day = static_cast<int>((index + day_offset) % 7);
                if (assignedToday[day].count(employee)) continue;
                const std::string& preferred = preferences.at(employee)[day];
                if (!CanAssign(employee, day, schedule, assignedToday, daysWorked)) continue;
                auto& slot = schedule[day][preferred];
                if (static_cast<int>(slot.size()) < PREFERRED_SHIFT_CAP)
                {
                    slot.push_back(employee);
                    assignedToday[day].insert(employee);
                    assignedThisRound.insert(employee);
                    daysWorked[employee]++;
                    break;
                }
                else
                {
                    std::vector<std::string> order;
                    if (priorityOrders)
                    {
                        auto iterator = priorityOrders->find(employee);
                        if (iterator != priorityOrders->end()) order = iterator->second;
                    }
                    if (order.empty())
                        for (const auto& shift_name : SHIFTS)
                            if (shift_name != preferred) order.push_back(shift_name);
                    bool assigned = false;
                    for (const auto& shift : order)
                    {
                        if (shift == preferred) continue;
                        if (!CanAssign(employee, day, schedule, assignedToday, daysWorked)) continue;
                        schedule[day][shift].push_back(employee);
                        assignedToday[day].insert(employee);
                        assignedThisRound.insert(employee);
                        daysWorked[employee]++;
                        assigned = true;
                        break;
                    }
                    if (assigned) break;
                }
            }
        }
    }

    // Second pass: ensure at least MIN_EMPLOYEES_PER_SHIFT per shift per day
    for (int day = 0; day < 7; ++day)
    {
        for (const auto& shift : SHIFTS)
        {
            int need = MIN_EMPLOYEES_PER_SHIFT - static_cast<int>(schedule[day][shift].size());
            if (need <= 0) continue;
            std::vector<std::string> candidates;
            for (const auto& employee : employees)
            {
                if (assignedToday[day].count(employee)) continue;
                if (daysWorked[employee] >= MAX_DAYS_PER_EMPLOYEE) continue;
                candidates.push_back(employee);
            }
            std::shuffle(candidates.begin(), candidates.end(), random_generator);
            for (const auto& employee : candidates)
            {
                if (need <= 0) break;
                schedule[day][shift].push_back(employee);
                assignedToday[day].insert(employee);
                daysWorked[employee]++;
                need--;
            }
        }
    }

    return schedule;
}

void PrintSchedule(const Schedule& schedule)
{
    std::cout << "\n===\n";
    std::cout << "WEEKLY EMPLOYEE SCHEDULE\n";
    std::cout << "===\n";
    for (int day = 0; day < 7; ++day)
    {
        std::cout << "\n" << DAYS[day] << ":\n";
        std::cout << "-----------------------\n";
        for (const auto& shift : SHIFTS)
        {
            const auto& names = schedule[day].at(shift);
            std::string capitalized = shift;
            if (!capitalized.empty()) capitalized[0] = static_cast<char>(std::toupper(static_cast<unsigned char>(capitalized[0])));
            std::cout << "  " << capitalized;
            for (size_t index = capitalized.size(); index < 12; ++index) std::cout << ' ';
            std::cout << " : ";
            if (names.empty())
            {
                std::cout << "(none)";
            }
            else
            {
                for (size_t index = 0; index < names.size(); ++index)
                {
                    if (index > 0) std::cout << ", ";
                    std::cout << names[index];
                }
            }
            std::cout << "\n";
        }
    }
    std::cout << "\n====\n";
}

/// <summary>
/// Run with sample data. This data was generated by ChatGPT by a prommpt of
/// "Generate 10 employees with their preferred shifts for a week with afternoon, morning, and evening shifts."
/// It was also provided with what kind of variable it would be stored like std::map<std::string, std::vector<std::string>>
/// </summary>
void RunDemoWithSampleData()
{
    std::map<std::string, std::vector<std::string>> preferences = {
        {"Alice", {"morning","morning","morning","morning","morning","morning","morning"}},
        {"Bob", {"afternoon","afternoon","afternoon","afternoon","afternoon","afternoon","afternoon"}},
        {"Carol", {"evening","evening","evening","evening","evening","evening","evening"}},
        {"Dave", {"morning","afternoon","evening","morning","afternoon","evening","morning"}},
        {"Eve", {"afternoon","evening","morning","afternoon","evening","morning","afternoon"}},
        {"Frank", {"morning","morning","afternoon","afternoon","evening","evening","morning"}},
        {"Grace", {"evening","morning","afternoon","evening","morning","afternoon","afternoon"}},
        {"Henry", {"afternoon","evening","morning","afternoon","evening","morning","evening"}},
        {"Ivy", {"morning","afternoon","evening","morning","afternoon","evening","morning"}},
        {"Jack", {"evening","morning","afternoon","evening","morning","afternoon","evening"}},
    };
    std::map<std::string, std::vector<std::string>> priorityOrders = {
        {"Alice", {"morning","afternoon","evening"}},
        {"Bob", {"afternoon","morning","evening"}},
        {"Carol", {"evening","afternoon","morning"}},
        {"Dave", {"morning","afternoon","evening"}},
        {"Eve", {"afternoon","evening","morning"}},
        {"Frank", {"morning","afternoon","evening"}},
        {"Grace", {"evening","morning","afternoon"}},
        {"Henry", {"afternoon","evening","morning"}},
        {"Ivy", {"morning","afternoon","evening"}},
        {"Jack", {"evening","morning","afternoon"}},
    };
    Schedule schedule = BuildSchedule(preferences, &priorityOrders);
    PrintSchedule(schedule);
}

}  // namespace

int main()
{
    std::cout << "Employee Schedule Manager\n";
    std::cout << "Options: 1 = Enter data manually  2 = Run demo with sample data  (type 'exit' anytime to quit)\n";
    std::cout << "Choice (1/2): ";
    std::string choice_str;
    std::cin >> choice_str;
    CheckExitCommand(choice_str);
    int choice = 0;
    try { choice = std::stoi(choice_str); } catch (...) {}
    if (choice == 2)
    {
        RunDemoWithSampleData();
        WaitForEnterBeforeExit();
        return 0;
    }
    std::cout << "Use priority ranking for shifts? (y/n): ";
    std::string use_priority_str;
    std::cin >> use_priority_str;
    CheckExitCommand(use_priority_str);
    bool use_priority = (!use_priority_str.empty() && (use_priority_str[0] == 'y' || use_priority_str[0] == 'Y'));

    std::map<std::string, std::vector<std::string>> preferences;
    std::map<std::string, std::vector<std::string>> priority_orders;
    CollectEmployeeData(preferences, use_priority ? &priority_orders : nullptr, use_priority);
    if (preferences.empty())
    {
        std::cout << "No employees entered. Exiting.\n";
        WaitForEnterBeforeExit();
        return 0;
    }
    Schedule schedule = BuildSchedule(preferences, use_priority ? &priority_orders : nullptr);
    PrintSchedule(schedule);
    WaitForEnterBeforeExit();
    return 0;
}

File: Advanced Programming Languages/Assignment3/employee_schedule.py
import random
import sys
from collections import defaultdict
from typing import Dict, List, Tuple, Optional

DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]


def CheckExitCommand(input_str: str) -> None:
    """If user typed 'exit' (case-insensitive), print message and exit the program."""
    if input_str.strip().lower() == "exit":
        print("Exiting.")
        sys.exit(0)
SHIFTS = ["morning", "afternoon", "evening"]
MIN_EMPLOYEES_PER_SHIFT = 2
MAX_DAYS_PER_EMPLOYEE = 5
MAX_ASSIGNMENT_PASSES = 3


def GetShiftPriorityInput() -> List[str]:
    """Get employee's shift preference order (1=first choice, 2=second, 3=third)."""
    print("  Enter priority order (1=most preferred, 2=second, 3=least preferred):")
    order = [None, None, None]
    for index in range(3):
        while True:
            rank_str = input(f"    {SHIFTS[index]}: rank (1-3): ")
            CheckExitCommand(rank_str)
            rank = int(rank_str)
            if 1 <= rank <= 3 and order[rank - 1] is None:
                order[rank - 1] = SHIFTS[index]
                break
            print("    Invalid or duplicate rank. Use 1, 2, 3 each once.")
    return order


def CollectEmployeeData(use_priority: bool = True) -> Tuple[Dict[str, List[str]], Optional[Dict[str, List[str]]]]:
    """
    Collect employee names and shift preferences for each day.
    Returns: (preferences[employee][day] = shift, optional priority_order[employee] = [shift1, shift2, shift3])
    """
    preferences: Dict[str, List[str]] = {}
    priority_orders: Optional[Dict[str, List[str]]] = {} if use_priority else None

    num_str = input("Number of employees: ")
    CheckExitCommand(num_str)
    number_of_employees = int(num_str)
    if number_of_employees <= 0:
        return preferences, priority_orders
    for _ in range(number_of_employees):
        name = input("Employee name: ").strip()
        CheckExitCommand(name)
        if not name:
            continue
        day_preferences = []
        if use_priority:
            priority = GetShiftPriorityInput()
            if priority_orders is not None:
                priority_orders[name] = priority
        for day in DAYS:
            while True:
                preference = input(f"  {day} preferred shift (morning/afternoon/evening): ").strip().lower()
                CheckExitCommand(preference)
                if preference in SHIFTS:
                    day_preferences.append(preference)
                    break
                print("    Invalid. Enter morning, afternoon, or evening.")
        preferences[name] = day_preferences

    return preferences, priority_orders


def BuildSchedule(
    preferences: Dict[str, List[str]],
    priority_orders: Optional[Dict[str, List[str]]],
) -> Dict[Tuple[int, str], List[str]]:
    """
    This scheduler favors employee preferences early, then relaxes constraints
    to guarantee minimum staffing coverage. Fairness is approximate, not optimal.
    """
    # schedule[(day, shift)] = list of employee names
    schedule: Dict[Tuple[int, str], List[str]] = defaultdict(list)
    days_worked: Dict[str, int] = defaultdict(int)
    assigned_today: Dict[int, set] = {day_index: set() for day_index in range(7)}

    employees = list(preferences.keys())
    random.shuffle(employees)  # fairness in tie-breaking

    def CanAssign(employee: str, day: int, shift: str) -> bool:
        if employee in assigned_today[day]:
            return False
        if days_worked[employee] >= MAX_DAYS_PER_EMPLOYEE:
            return False
        return True

    # First pass: assign by preference in rounds; at most one day per employee per round.
    # Use MAX_ASSIGNMENT_PASSES rounds so backfill has enough capacity to ensure at least 2 per shift on all days.
    for round_index in range(MAX_ASSIGNMENT_PASSES):
        assigned_this_round: set = set()
        for index, employee in enumerate(employees):
            if employee in assigned_this_round or days_worked[employee] >= MAX_DAYS_PER_EMPLOYEE:
                continue
            for day_offset in range(7):
                day = (index + day_offset) % 7
                if employee in assigned_today[day]:
                    continue
                preferred = preferences[employee][day]
                if not CanAssign(employee, day, preferred):
                    continue
                if len(schedule[(day, preferred)]) < 3:
                    schedule[(day, preferred)].append(employee)
                    assigned_today[day].add(employee)
                    assigned_this_round.add(employee)
                    days_worked[employee] += 1
                    break
                else:
                    # Conflict: assign to another shift same day (use priority order if available)
                    if priority_orders:
                        order = priority_orders.get(employee, [])
                    else:
                        order = []
                    if not order:
                        order = [preferred] + [s for s in SHIFTS if s != preferred]
                    for shift_option in order:
                        if shift_option != preferred and CanAssign(employee, day, shift_option):
                            schedule[(day, shift_option)].append(employee)
                            assigned_today[day].add(employee)
                            assigned_this_round.add(employee)
                            days_worked[employee] += 1
                            break
                    else:
                        continue
                    break

    # Second pass: ensure at least MIN_EMPLOYEES_PER_SHIFT per shift per day
    for day in range(7):
        for shift in SHIFTS:
            need = MIN_EMPLOYEES_PER_SHIFT - len(schedule[(day, shift)])
            if need <= 0:
                continue
            candidates = [
                employee for employee in employees
                if employee not in assigned_today[day] and days_worked[employee] < MAX_DAYS_PER_EMPLOYEE
            ]
            random.shuffle(candidates)
            for employee in candidates:
                if need <= 0:
                    break
                schedule[(day, shift)].append(employee)
                assigned_today[day].add(employee)
                days_worked[employee] += 1
                need -= 1

    return schedule


def PrintSchedule(schedule: Dict[Tuple[int, str], List[str]]) -> None:
    """Output the final weekly schedule in a readable format."""
    print("\n" + "=" * 70)
    print("WEEKLY EMPLOYEE SCHEDULE")
    print("=" * 70)
    for day in range(7):
        print(f"\n{DAYS[day]}:")
        print("-" * 40)
        for shift in SHIFTS:
            names = schedule[(day, shift)]
            print(f"  {shift.capitalize():12} : {', '.join(names) if names else '(none)'}")
    print("\n" + "=" * 70)


def RunDemoWithSampleData() -> None:
    """Run with built-in sample data (no user input) for quick testing."""
    preferences = {
        "Alice": ["morning"] * 7,
        "Bob": ["afternoon"] * 7,
        "Carol": ["evening"] * 7,
        "Dave": ["morning", "afternoon", "evening", "morning", "afternoon", "evening", "morning"],
        "Eve": ["afternoon", "evening", "morning", "afternoon", "evening", "morning", "afternoon"],
        "Frank": ["morning", "morning", "afternoon", "afternoon", "evening", "evening", "morning"],
        "Grace": ["evening", "morning", "afternoon", "evening", "morning", "afternoon", "afternoon"],
        "Henry": ["afternoon", "evening", "morning", "afternoon", "evening", "morning", "evening"],
        "Ivy": ["morning", "afternoon", "evening", "morning", "afternoon", "evening", "morning"],
        "Jack": ["evening", "morning", "afternoon", "evening", "morning", "afternoon", "evening"],
    }
    priority_orders = {
        "Alice": ["morning", "afternoon", "evening"],
        "Bob": ["afternoon", "morning", "evening"],
        "Carol": ["evening", "afternoon", "morning"],
        "Dave": ["morning", "afternoon", "evening"],
        "Eve": ["afternoon", "evening", "morning"],
        "Frank": ["morning", "afternoon", "evening"],
        "Grace": ["evening", "morning", "afternoon"],
        "Henry": ["afternoon", "evening", "morning"],
        "Ivy": ["morning", "afternoon", "evening"],
        "Jack": ["evening", "morning", "afternoon"],
    }
    schedule = BuildSchedule(preferences, priority_orders)
    PrintSchedule(schedule)


def Main() -> None:
    print("Employee Schedule Manager")
    print("Options: 1 = Enter data manually  2 = Run demo with sample data  (type 'exit' anytime to quit)")
    choice = input("Choice (1/2): ").strip()
    CheckExitCommand(choice)
    if choice == "2":
        RunDemoWithSampleData()
        return
    use_priority_str = input("Use priority ranking for shifts? (y/n): ").strip()
    CheckExitCommand(use_priority_str)
    use_priority = use_priority_str.lower() == "y"
    preferences, priority_orders = CollectEmployeeData(use_priority)
    if not preferences:
        print("No employees entered. Exiting.")
        return
    schedule = BuildSchedule(preferences, priority_orders)
    PrintSchedule(schedule)


if __name__ == "__main__":
    Main()
